packages/babel-plugin-proposal-destructuring-private/src/index.ts

```diff
@@ -0,0 +1,213 @@
+import { declare } from "@babel/helper-plugin-utils";
+import syntaxDestructuringPrivate from "@babel/plugin-syntax-destructuring-private";
+import {
+  hasPrivateKeys,
+  hasPrivateClassElement,
+  transformPrivateKeyDestructuring,
+  buildVariableDeclarationFromParams,
+} from "./util";
+import { convertFunctionParams } from "@babel/plugin-transform-parameters";
+import { unshiftForXStatementBody } from "@babel/plugin-transform-destructuring";
+
+import type { PluginPass } from "@babel/core";
+import type { Visitor } from "@babel/traverse";
+
+export default declare(function ({ assertVersion, assumption, types: t }) {
+  assertVersion("^7.17.0");
+  const {
+    assignmentExpression,
+    assignmentPattern,
+    cloneNode,
+    expressionStatement,
+    isExpressionStatement,
+    isIdentifier,
+    isSequenceExpression,
+    sequenceExpression,
+    variableDeclaration,
+    variableDeclarator,
+  } = t;
+
+  const ignoreFunctionLength = assumption("ignoreFunctionLength");
+  const objectRestNoSymbols = assumption("objectRestNoSymbols");
+
+  const privateKeyDestructuringVisitor: Visitor<PluginPass> = {
+    Function(path) {
+      // (b, { #x: x } = I) => body
+      // transforms to:
+      // (b, p1) => { var { #x: x } = p1 === undefined ? I : p1; body; }
+      const firstPrivateIndex = path.node.params.findIndex(param =>
+        hasPrivateKeys(param),
+      );
+      if (firstPrivateIndex === -1) return;
+      // wrap function body within IIFE if any param is shadowed
+      convertFunctionParams(path, ignoreFunctionLength, () => false, false);
+      // invariant: path.body is always a BlockStatement after `convertFunctionParams`
+      const { node, scope } = path;
+      const params = node.params;
+      const firstAssignmentPatternIndex = ignoreFunctionLength
+        ? -1
+        : params.findIndex(param => param.type === "AssignmentPattern");
+      const paramsAfterIndex = params.splice(firstPrivateIndex);
+      const { params: transformedParams, variableDeclaration } =
+        buildVariableDeclarationFromParams(paramsAfterIndex, scope);
+
+      path.get("body").unshiftContainer("body", variableDeclaration);
+      params.push(...transformedParams);
+      // preserve function.length
+      // (b, p1) => {}
+      // transforms to
+      // (b, p1 = void 0) => {}
+      if (firstAssignmentPatternIndex >= firstPrivateIndex) {
```

Comment:
We can not assign the `firstAssignmentPatternIndex` here because the transform will convert any params after the first private destructuring. In other words, `firstAssignmentPatternIndex` will always be less than `firstPrivateIndex` if we assign `firstAssignmentPatternIndex` here.